src/account/account.cairo

```diff
@@ -0,0 +1,218 @@
+    // IERC721Metadata
+    fn name(self: @TContractState) -> felt252;
+    fn symbol(self: @TContractState) -> felt252;
+    fn token_uri(self: @TContractState, token_id: u256) -> felt252;
+}
+
+#[starknet::contract]
+mod Account {
+
+use starknet::{get_tx_info, get_caller_address, get_contract_address};
+use ecdsa::check_ecdsa_signature;
+use array::SpanTrait;
+use array::ArrayTrait;
+use box::BoxTrait;
+use option::OptionTrait;
+use zeroable::Zeroable;
+use starknet::account::Call;
+use starknet::ContractAddress;
+use super::IERC721DispatcherTrait;
+use super::IERC721Dispatcher;
+
+    #[storage]
+    struct Storage{
+        _public_key: felt252,
+        _token_contract:ContractAddress,
+        _token_id:u256,
+    }
+
+
+
+        #[constructor]
+        fn constructor(ref self: ContractState, _public_key:felt252, token_contract:ContractAddress, token_id:u256){
+        self.initializer(_public_key, token_contract, token_id);
+        }
+
+
+        #[external(v0)]
+        impl IAccountImpl of super::IAccount<ContractState>{
+            fn get_public_key(self: @ContractState) -> felt252{
+                self._public_key.read()
+            }
+
+            fn set_public_key(ref self:ContractState, new_public_key:felt252){
+                assert_only_self();
+                self._public_key.write(new_public_key);
+            }
+
+            fn __validate_deploy__(
+                self: @ContractState,
+                class_hash:felt252,
+                contract_address_salt:felt252,
+                public_key:felt252
+            ) -> felt252{
+                self.validate_transaction()
+            }
+
+            fn __validate_declare__(self: @ContractState, class_hash: felt252) -> felt252 {
+                        self.validate_transaction()
+            }
+
+            fn is_valid_signature(
+                        self: @ContractState, hash: felt252, signature: Array<felt252>
+                    ) -> felt252 {
+                        if self._is_valid_signature(hash, signature.span()) {
+                            starknet::VALIDATED
+                        } else {
+                            0
+                        }
+                    }
+
+                fn __validate__( ref self: ContractState, mut calls:Array<Call>) -> felt252{
+                    self.validate_transaction()
+                }
+                // equivalent of executeCall
+                fn __execute__(ref self: ContractState, mut calls:Array<Call>, token_contract:ContractAddress, token_id:u256) -> Array<Span<felt252>>{
+                let owner = IERC721Dispatcher { contract_address: token_contract }.owner_of(token_id);
+                assert(owner == get_caller_address(), 'CALLER_IS_NOT_OWNER');
+                    let sender = get_caller_address();
+                    assert(sender.is_zero(), 'Account: invalid caller');
```

Comment:
This assert statement contradicts the first `assert(owner == get_caller_address(), 'CALLER_IS_NOT_OWNER');`

What do we check for exactly? checking that the caller is zero helps prevent reentrancy, but how do we enforce the `execute` function can be called by only the NFT owner of an account? 

I think we could remove the first assert, because we definitely want to check that caller is a zero address, but `is_valid_signature` would need to assert signature is that of NFT owner, which means the NFT owner, would need to ensure the supplied public key in constructor should be his.